File: .codebuddy/skills/format-validation/scripts/validate_json.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"strings"
)

// ValidationResult represents the result of validating a single file
type ValidationResult struct {
	FilePath string
	Valid    bool
	Error    string
}

// ValidationSummary contains summary statistics
type ValidationSummary struct {
	Directory      string
	TotalFiles     int
	ValidFiles     int
	InvalidFiles   int
	InvalidDetails []ValidationResult
}

func main() {
	dir := flag.String("dir", "", "Directory to scan for JSON files")
	noRecursive := flag.Bool("no-recursive", false, "Scan only top-level directory")
	flag.Parse()

	if *dir == "" {
		fmt.Println("Usage: go run validate_json.go -dir <directory> [-no-recursive]")
		fmt.Println("\nOptions:")
		fmt.Println("  -dir           Path to directory to scan for JSON files")
		fmt.Println("  -no-recursive  Scan only the top-level directory (default: recursive)")
		os.Exit(1)
	}

	fmt.Printf("Validating JSON files in: %s\n", *dir)
	if *noRecursive {
		fmt.Println("(top-level only)")
	} else {
		fmt.Println("(recursive scan)")
	}
	fmt.Println()

	summary := validateJSONFiles(*dir, !*noRecursive)
	printSummary(summary)

	if summary.InvalidFiles > 0 {
		os.Exit(1)
	}
}

// validateJSONFiles validates all JSON files in the specified directory
func validateJSONFiles(directory string, recursive bool) ValidationSummary {
	summary := ValidationSummary{
		Directory: directory,
	}

	var jsonFiles []string
	var err error

	if recursive {
		jsonFiles, err = findJSONFilesRecursive(directory)
	} else {
		jsonFiles, err = findJSONFiles(directory)
	}

	if err != nil {
		fmt.Printf("Error scanning directory: %v\n", err)
		return summary
	}

	summary.TotalFiles = len(jsonFiles)

	for _, filePath := range jsonFiles {
		result := validateJSONFile(filePath)
		if result.Valid {
			summary.ValidFiles++
			fmt.Printf("✓ %s\n", filePath)
		} else {
			summary.InvalidFiles++
			summary.InvalidDetails = append(summary.InvalidDetails, result)
			fmt.Printf("✗ %s\n", filePath)
			fmt.Printf("  Error: %s\n", result.Error)
		}
	}

	return summary
}

// findJSONFilesRecursive finds all JSON files in directory and subdirectories
func findJSONFilesRecursive(root string) ([]string, error) {
	var jsonFiles []string

	err := filepath.WalkDir(root, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}

		if !d.IsDir() && strings.HasSuffix(strings.ToLower(path), ".json") {
			jsonFiles = append(jsonFiles, path)
		}

		return nil
	})

	return jsonFiles, err
}

// findJSONFiles finds JSON files only in the top-level directory
func findJSONFiles(directory string) ([]string, error) {
	entries, err := os.ReadDir(directory)
	if err != nil {
		return nil, err
	}

	var jsonFiles []string
	for _, entry := range entries {
		if !entry.IsDir() && strings.HasSuffix(strings.ToLower(entry.Name()), ".json") {
			jsonFiles = append(jsonFiles, filepath.Join(directory, entry.Name()))
		}
	}

	return jsonFiles, nil
}

// validateJSONFile validates a single JSON file
func validateJSONFile(filePath string) ValidationResult {
	result := ValidationResult{
		FilePath: filePath,
		Valid:    true,
	}

	data, err := os.ReadFile(filePath)
	if err != nil {
		result.Valid = false
		result.Error = fmt.Sprintf("Error reading file: %v", err)
		return result
	}

	var v interface{}
	if err := json.Unmarshal(data, &v); err != nil {
		result.Valid = false
		result.Error = fmt.Sprintf("JSON decode error: %v", err)
		return result
	}

	return result
}

// printSummary prints the validation summary
func printSummary(summary ValidationSummary) {
	fmt.Println()
	fmt.Println("============================================================")
	fmt.Println("VALIDATION SUMMARY")
	fmt.Println("============================================================")
	fmt.Printf("Directory: %s\n", summary.Directory)
	fmt.Printf("Total JSON files: %d\n", summary.TotalFiles)
	fmt.Printf("Valid files: %d\n", summary.ValidFiles)
	fmt.Printf("Invalid files: %d\n", summary.InvalidFiles)

	if summary.InvalidFiles > 0 {
		fmt.Println()
		fmt.Println("Invalid files:")
		for _, detail := range summary.InvalidDetails {
			fmt.Printf("  - %s\n", detail.FilePath)
			fmt.Printf("    %s\n", detail.Error)
		}
	}

	fmt.Println("============================================================")
}

File: .codebuddy/skills/format-validation/scripts/validate_json.py
#!/usr/bin/env python3
"""
JSON Format Validation Script
Scans directories and validates JSON file formats
"""

import json
import os
import sys
from pathlib import Path
from typing import Dict, List, Tuple


def validate_json_file(file_path: str) -> Tuple[bool, str]:
    """
    Validate a single JSON file

    Args:
        file_path: Path to the JSON file

    Returns:
        Tuple of (is_valid, error_message)
    """
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            json.load(f)
        return True, ""
    except json.JSONDecodeError as e:
        return False, f"JSON decode error at line {e.lineno}, column {e.colno}: {e.msg}"
    except Exception as e:
        return False, f"Error reading file: {str(e)}"


def scan_directory(directory: str, recursive: bool = True) -> List[str]:
    """
    Scan directory for JSON files

    Args:
        directory: Directory to scan
        recursive: Whether to scan subdirectories

    Returns:
        List of JSON file paths
    """
    json_files = []
    path = Path(directory)

    if not path.exists():
        print(f"Error: Directory '{directory}' does not exist")
        return []

    if not path.is_dir():
        print(f"Error: '{directory}' is not a directory")
        return []

    if recursive:
        pattern = '**/*.json'
    else:
        pattern = '*.json'

    for file_path in path.glob(pattern):
        # Skip files/directories in hidden folders (starting with .)
        if any(part.startswith('.') for part in file_path.parts):
            continue
        # Skip files/directories in venv, __pycache__, node_modules, etc.
        ignored_dirs = ['venv', '__pycache__', 'node_modules', '.venv', 'env', 'virtualenv']
        if any(part in ignored_dirs for part in file_path.parts):
            continue
        if file_path.is_file():
            json_files.append(str(file_path))

    return sorted(json_files)


def validate_json_files(directory: str, recursive: bool = True) -> Dict:
    """
    Validate all JSON files in directory

    Args:
        directory: Directory to scan
        recursive: Whether to scan subdirectories

    Returns:
        Dictionary with validation results
    """
    json_files = scan_directory(directory, recursive)
    results = {
        'directory': directory,
        'total_files': len(json_files),
        'valid_files': 0,
        'invalid_files': 0,
        'valid': [],
        'invalid': []
    }

    for file_path in json_files:
        is_valid, error_msg = validate_json_file(file_path)

        if is_valid:
            results['valid_files'] += 1
            results['valid'].append(file_path)
            print(f"✓ {file_path}")
        else:
            results['invalid_files'] += 1
            results['invalid'].append({
                'file': file_path,
                'error': error_msg
            })
            print(f"✗ {file_path}")
            print(f"  Error: {error_msg}")

    return results


def print_summary(results: Dict):
    """Print validation summary"""
    print("\n" + "="*60)
    print("VALIDATION SUMMARY")
    print("="*60)
    print(f"Directory: {results['directory']}")
    print(f"Total JSON files: {results['total_files']}")
    print(f"Valid files: {results['valid_files']}")
    print(f"Invalid files: {results['invalid_files']}")

    if results['invalid_files'] > 0:
        print("\nInvalid files:")
        for item in results['invalid']:
            print(f"  - {item['file']}")
            print(f"    {item['error']}")

    print("="*60)


def main():
    """Main entry point"""
    if len(sys.argv) < 2:
        print("Usage: python validate_json.py <directory> [--no-recursive]")
        print("\nArguments:")
        print("  directory    Path to directory to scan for JSON files")
        print("  --no-recursive  Scan only the top-level directory (default: recursive)")
        sys.exit(1)

    directory = sys.argv[1]
    recursive = '--no-recursive' not in sys.argv[2:]

    print(f"Validating JSON files in: {directory}")
    if recursive:
        print("(recursive scan)")
    else:
        print("(top-level only)")
    print()

    results = validate_json_files(directory, recursive)
    print_summary(results)

    # Exit with error code if any files are invalid
    sys.exit(1 if results['invalid_files'] > 0 else 0)


if __name__ == '__main__':
    main()
